fetcher/fetcher.go

```diff
@@ -79,144 +82,147 @@ func (f *Fetcher) runWorker() {
-		if f.keysToFetch[k].cache != nil {
+		d, ok := f.keys[k]
+		if !ok {
+			f.keys[k] = &key{blocked: []ids.ID{txID}}
+			tasks = append(tasks, &task{
+				ctx: ctx,
+				id:  txID,
+				key: k,
+			})
 			continue
 		}
-		f.keysToFetch[k].queue = append(f.keysToFetch[k].queue, txID)
-		t := &task{
-			ctx: ctx,
-			id:  txID,
-			key: k,
+
+		// Don't register that we are blocked if the key has already been fetched
+		if d.cache != nil {
+			continue
 		}
-		tasks = append(tasks, t)
+
+		// Register to get notified when the key is fetched
+		d.blocked = append(d.blocked, txID)
 	}
+	wg := &sync.WaitGroup{}
 	wg.Add(len(tasks))
-	f.txnsToFetch[txID] = wg
-	f.fetchLock.Unlock()
+	f.txs[txID] = &tx{wg, stateKeys}
+	f.l.Unlock()
 
+	// Send fetch tasks to the workers
 	for _, t := range tasks {
-		f.fetchable <- t
+		f.tasks <- t
 	}
-	return wg
+	return nil
 }
 
-func (f *Fetcher) Get(wg *sync.WaitGroup, stateKeys state.Keys) (map[string]uint16, map[string][]byte) {
-	// Block until all keys for the tx are fetched
-	wg.Wait()
-
-	f.l.Lock()
-	f.completed++
-	f.l.Unlock()
-
-	f.fetchLock.RLock()
-	defer f.fetchLock.RUnlock()
+// Get will return the state keys fetched for the txID or return the error
+// that caused fetching to fail.
+//
+// Get can be called concurrently.
+func (f *Fetcher) Get(txID ids.ID) (map[string]uint16, map[string][]byte, error) {
+	// Block until all keys for the tx are fetched or if the fetcher errored
+	f.l.RLock()
+	tx, ok := f.txs[txID]
+	fErr := f.err // allows us to avoid deadlock when [tx.wg] will never be cleared
+	f.l.RUnlock()
+	if !ok {
+		return nil, nil, fmt.Errorf("%w: unable to get keys for transaction %s", ErrMissingTx, txID)
+	}
+	if fErr != nil {
+		return nil, nil, fErr
+	}
+	tx.wg.Wait()
 
 	// Fetch keys from cache
+	f.l.RLock()
+	defer f.l.RUnlock()
 	var (
-		reads   = make(map[string]uint16, len(stateKeys))
-		storage = make(map[string][]byte, len(stateKeys))
+		stateKeys = tx.keys
+		reads     = make(map[string]uint16, len(stateKeys))
+		storage   = make(map[string][]byte, len(stateKeys))
 	)
 	for k := range stateKeys {
-		if v := f.keysToFetch[k].cache; v != nil {
+		if v := f.keys[k].cache; v != nil {
 			reads[k] = v.chunks
 			if v.exists {
 				storage[k] = v.v
 			}
 		}
 	}
-	return reads, storage
+	return reads, storage, nil
 }
 
+// Stop terminates the fetcher before all keys have been fetched.
 func (f *Fetcher) Stop() {
-	f.stopOnce.Do(func() {
-		f.err = ErrStopped
-		close(f.stop)
-	})
+	f.handleErr(ErrStopped)
 }
 
-// Wait until all the workers are done and return any errors
+// Wait until all the workers are done and return any errors.
+//
+// [Wait] can be called multiple times, however, [Fetch] should never be
+// called after [Wait] is called.
 func (f *Fetcher) Wait() error {
-	f.l.Lock()
-	if f.completed == f.numTxs {
-		close(f.fetchable)
-	}
-	f.l.Unlock()
+	f.waitOnce.Do(func() {
+		close(f.tasks)
+	})
 	f.wg.Wait()
 	return f.err
```

Comment:
We don't need a lock around `f.err` because it can never be updated after `f.wg.Wait()` returns.